File: python/tests/test_sst.py
"""SST / SST* (Li, Littlefield & Bekris 2016): forward-propagation kinodynamic
planner that reaches the goal on open space, fails when walled off, and keeps the
active tree sparse via the witness set (which SST* shrinks over iterations)."""

from __future__ import annotations

from pathlib import Path
from typing import Any

import pytest
import yaml
from conftest import CONFIG_DIR, grid_from, open_grid

from navigation.core.params import ParamError, ParamSet
from navigation.global_planning.sampling.sst import SST
from navigation.maps.occupancy_grid import OccupancyGrid2D

_ALGO = "sst"


def _config(tmp_path: Path, **overrides: Any) -> ParamSet:
    """Real config with selected defaults overridden (smaller budgets keep tests fast)."""
    doc = yaml.safe_load((CONFIG_DIR / f"{_ALGO}.yaml").read_text())
    for entry in doc["params"]:
        if entry["name"] in overrides:
            entry["default"] = overrides[entry["name"]]
    out = tmp_path / f"{_ALGO}.yaml"
    out.write_text(yaml.safe_dump(doc), encoding="utf-8")
    return ParamSet.from_yaml(out)


def _collision_free(grid: OccupancyGrid2D, path: list[tuple[float, float]]) -> bool:
    return all(grid.is_state_valid(p) for p in path) and all(
        grid.is_motion_valid(a, b) for a, b in zip(path, path[1:], strict=False)
    )


def test_finds_dynamically_feasible_path(tmp_path: Path) -> None:
    grid = open_grid(10, 10, seed=1)
    start, goal = (0.5, 0.5), (8.5, 8.5)
    params = _config(tmp_path, max_iterations=20000)
    result = SST(params).plan(grid, start, goal)
    assert result.success
    assert result.path[0] == pytest.approx(start)
    # Goal reached within tolerance (position-only; the unicycle has no goal heading).
    assert grid.distance(result.path[-1], goal) <= params.get_float("goal_tolerance")
    assert _collision_free(grid, result.path)
    assert result.cost > 0.0
    # Dynamic feasibility: the path is a densely-sampled propagated trajectory, not
    # straight steer jumps — every consecutive chord stays within the arc spacing.
    max_step = max(grid.distance(a, b) for a, b in zip(result.path, result.path[1:], strict=False))
    assert max_step <= 0.25


def test_no_path_when_walled_off(tmp_path: Path) -> None:
    # Fully occupied middle column: no propagated arc can cross it.
    grid = grid_from(["..#..", "..#..", "..#..", "..#..", "..#.."], seed=3)
    params = _config(tmp_path, max_iterations=500)
    result = SST(params).plan(grid, (0.5, 0.5), (4.5, 4.5))
    assert not result.success
    assert result.path == []
    assert result.cost == 0.0


def test_out_of_range_default_rejected(tmp_path: Path) -> None:
    # goal_bias default above its declared max must fail validation at load time.
    doc = yaml.safe_load((CONFIG_DIR / f"{_ALGO}.yaml").read_text())
    for entry in doc["params"]:
        if entry["name"] == "goal_bias":
            entry["default"] = 2.0
    bad = tmp_path / f"{_ALGO}.yaml"
    bad.write_text(yaml.safe_dump(doc), encoding="utf-8")
    with pytest.raises(ParamError):
        ParamSet.from_yaml(bad)


def test_witness_set_bounds_active_growth(tmp_path: Path) -> None:
    # The witness set caps how many nodes stay active: over a large budget the active
    # tree stays sparse (far below the iteration count), while total propagations added
    # exceed it because dominated representatives are pruned. A naive RRT would keep
    # every added node, so its tree would grow ~linearly in iterations instead.
    grid = open_grid(12, 12, seed=2)
    result = SST(_config(tmp_path, max_iterations=20000)).plan(grid, (0.5, 0.5), (10.5, 10.5))
    assert result.success
    assert result.stats.tree_size < result.stats.iterations // 10  # sparse active set
    assert result.stats.expanded_nodes > result.stats.tree_size  # pruning happened
    assert result.stats.tree_size <= 800


def test_sst_star_shrinking_schedule_runs(tmp_path: Path) -> None:
    # SST* shrinks delta_bn / delta_s over iterations; it must run end-to-end without
    # error and grow a tree (finer sparsification keeps more active nodes than SST).
    grid = open_grid(10, 10, seed=1)
    result = SST(_config(tmp_path, sst_star=True, max_iterations=4000)).plan(
        grid, (0.5, 0.5), (8.5, 8.5)
    )
    assert isinstance(result.success, bool)
    assert result.stats.iterations == 4000
    assert result.stats.tree_size > 1

File: cpp/tests/test_sst.cpp
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "navigation/core/params.hpp"
#include "navigation/global_planning/sampling/sst.hpp"
#include "test_util.hpp"

using namespace navigation;
using core::Point;

namespace {

// Open square grid (resolution 0.5): '.' everywhere.
std::vector<std::string> open_rows(int n) {
  return std::vector<std::string>(static_cast<size_t>(n), std::string(static_cast<size_t>(n), '.'));
}

// Full SST config; smaller iteration budgets keep the tests fast.
std::string sst_cfg(int max_iterations, const std::string& sst_star = "false") {
  return test::write_temp(
      "sst.yaml",
      "algorithm: sst\ncategory: global_planning\nparams:\n"
      "  - {name: max_iterations, type: int, default: " + std::to_string(max_iterations) +
          ", min: 1, max: 2000000, description: n}\n"
      "  - {name: goal_bias, type: float, default: 0.1, min: 0.0, max: 1.0, description: b}\n"
      "  - {name: goal_tolerance, type: float, default: 0.6, min: 0.0, max: 100, description: t}\n"
      "  - {name: delta_bn, type: float, default: 1.2, min: 0.01, max: 100, description: bn}\n"
      "  - {name: delta_s, type: float, default: 0.5, min: 0.01, max: 100, description: ds}\n"
      "  - {name: max_velocity, type: float, default: 1.5, min: 0.01, max: 100, description: v}\n"
      "  - {name: max_omega, type: float, default: 1.5, min: 0.0, max: 100, description: w}\n"
      "  - {name: prop_duration_min, type: float, default: 0.2, min: 0.001, max: 100, description: pmn}\n"
      "  - {name: prop_duration_max, type: float, default: 0.8, min: 0.001, max: 100, description: pmx}\n"
      "  - {name: sst_star, type: bool, default: " + sst_star + ", description: star}\n"
      "  - {name: seed, type: int, default: 1, min: 0, max: 2147483647, description: s}\n");
}

}  // namespace

// (a) valid, collision-free, dynamically-feasible path on an open grid -------------

TEST(Sst, FindsDynamicallyFeasiblePathOnOpenGrid) {
  auto g = test::make_grid(open_rows(20));  // 10m x 10m world
  Point start{0.5, 0.5}, goal{9.0, 9.0};
  global_planning::SstPlanner p(core::ParamSet::from_yaml(sst_cfg(30000)));
  auto r = p.plan(g, start, goal, nullptr);
  ASSERT_TRUE(r.success);
  // Goal reached within tolerance (position-only; the unicycle has no goal heading).
  EXPECT_LE(g.distance(r.path.back(), goal), 0.6);
  double max_step = 0.0;
  for (size_t i = 0; i + 1 < r.path.size(); ++i) {
    EXPECT_TRUE(g.is_state_valid(r.path[i]));
    EXPECT_TRUE(g.is_motion_valid(r.path[i], r.path[i + 1])) << "arc chord must be collision-free";
    max_step = std::max(max_step, g.distance(r.path[i], r.path[i + 1]));
  }
  // Dynamic feasibility: a densely-sampled propagated trajectory, not steer jumps.
  EXPECT_LE(max_step, 0.25);
}

// (b) no-path case: a full wall separates start from goal --------------------------

TEST(Sst, NoPathAcrossFullWall) {
  auto rows = open_rows(10);
  rows[5] = "##########";  // impassable horizontal wall
  auto g = test::make_grid(rows);
  Point start{0.25, 0.25}, goal{4.75, 4.75};
  global_planning::SstPlanner p(core::ParamSet::from_yaml(sst_cfg(500)));
  auto r = p.plan(g, start, goal, nullptr);
  EXPECT_FALSE(r.success);
  EXPECT_TRUE(r.path.empty());
}

// (c) param validation failure: out-of-range default ------------------------------

TEST(Sst, BadGoalBiasThrows) {
  std::string bad = test::write_temp(
      "sst.yaml",
      "algorithm: sst\ncategory: global_planning\nparams:\n"
      "  - {name: goal_bias, type: float, default: 2.0, min: 0.0, max: 1.0, description: over max}\n");
  EXPECT_THROW(core::ParamSet::from_yaml(bad), std::runtime_error);
}

// (d) witness set bounds active-node growth; SST* schedule runs --------------------

TEST(Sst, WitnessSetBoundsActiveGrowth) {
  auto g = test::make_grid(open_rows(24));  // 12m x 12m world
  Point start{0.5, 0.5}, goal{11.0, 11.0};
  global_planning::SstPlanner p(core::ParamSet::from_yaml(sst_cfg(20000)));
  auto r = p.plan(g, start, goal, nullptr);
  ASSERT_TRUE(r.success);
  EXPECT_LT(r.stats.tree_size, r.stats.iterations / 10);   // sparse active set
  EXPECT_GT(r.stats.expanded_nodes, r.stats.tree_size);    // pruning happened
  EXPECT_LE(r.stats.tree_size, 800);
}

TEST(Sst, SstStarShrinkingScheduleRuns) {
  auto g = test::make_grid(open_rows(20));
  Point start{0.5, 0.5}, goal{9.0, 9.0};
  global_planning::SstPlanner p(core::ParamSet::from_yaml(sst_cfg(4000, "true")));
  auto r = p.plan(g, start, goal, nullptr);  // must not crash
  EXPECT_EQ(r.stats.iterations, 4000);
  EXPECT_GT(r.stats.tree_size, 1);
}
